### PR TITLE
[MXNET-683] [WIP] Adding Inplace Activated Batch Normalization

### PR DESCRIPTION
**Pull request by zhanghang1989**
## Description ##

Adding Inplace Activated Batch Normalization

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] mx.nd.contrib.InplaceABN
- [x] mx.gluon.contrib.nn.InplaceABN

### CODE REVIEW

File: tests/python/gpu/test_operator_gpu.py [hunk 1]
```diff
@@ -1996,6 +1996,31 @@ def get_num_devices():
...
+        output2 = layer2(input)
+        loss1 = (output1 ** 2).sum()
+        loss2 = (output2 ** 2).sum()
+        mx.autograd.backward(loss1)
+        mx.autograd.backward(loss2)
+    assert_almost_equal(loss1.asnumpy(), loss2.asnumpy(), atol=1e-3, rtol=1e-3)
+    assert_almost_equal(output1.asnumpy(), output2.asnumpy(), atol=1e-3, rtol=1e-3)
+    assert_almost_equal(input1.grad.asnumpy(), input2.grad.asnumpy(), atol=1e-3, rtol=1e-3)
+
+def test_inpabn():
```
Comment (szha) [inline tests/python/gpu/test_operator_gpu.py]: This doesn't seem the right place to put this test. For this PR, it would be better to have:
1. An operator test in https://github.com/apache/incubator-mxnet/blob/master/tests/python/unittest/test_contrib_operator.py
2. A gluon block test in https://github.com/apache/incubator-mxnet/blob/master/tests/python/unittest/test_gluon_contrib.py

Comment (zhanghang1989) [reply tests/python/gpu/test_operator_gpu.py]: This will be a gpu test. Still working in progress.

Comment (szha) [reply tests/python/gpu/test_operator_gpu.py]: I see. Nonetheless, given the scope of the change the suggested tests should be appropriate.

**Comment by stu1130** [comment]
@zhanghang1989 could you please address the cpplint issue. Thanks

**Comment by zhanghang1989** [comment]
Thanks @stu1130 . The implementation is not fully working. Though it can pass the unit test and can even train for a few epochs. But it will fail in the middle of the training. I will debug it later when I have more time.

**Comment by vandanavk** [comment]
@sandeep-krishnamurthy Please change label to pr-work-in-progress
